### PR TITLE
Change static methods to class methods

### LINKED ISSUE #1695: Classmethods instead of staticmethods
I noticed that many methods of the Tensor class are marked as static methods, even though they return new Tensor instances. These methods seem more like class methods than static methods. Converting them to class methods would:

1. Make it more clear that those methods create new Tensors
2. Will support subclassing Tensor

Is there any reason they are currently static methods? If there are no arguments against changing those static methods to class methods, I can create a PR for it.

### PR DESCRIPTION
**Pull request by yochem**
Static methods that use their own class, e.g. to create a new instance of its own class should be class methods. This makes the methods also work on subclasses.

Closes #1695.

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/lazy.py                           3               3
tinygrad/tensor.py                        54              55
------------------------------------------------------------
total                                     57              58
------------------------------------------------------------
lines added in the tinygrad folder: -1
```

**Comment by yochem** [comment]
I also want to type-hint the class methods. In Python 3.11, this is done using [`typing.Self`](https://docs.python.org/3/library/typing.html#typing.Self). <3.11, this can be done by just specifying the class name:

e.g. in Tensor:
```python
def eye(cls:Tensor, dim:int, **kwargs) -> Tensor:
    ...
```

But this doesn't work for subclasses, as they return the subclass of Tensor, and not Tensor. So this is the recommended way for type hinting `cls` and the return value of the class method:

```python
T = TypeVar('T', bound=Tensor)
def eye(cls:T, dim:int, **kwargs) -> T:
    ...
```

This however (1) adds a line to the codebase and (2) is probably not very clear to all users what it does. We could of course use a better name than `T`, but still.

What do you think? Leave the type hints for now, or go with one of the two options above?

**Comment by roelofvandijk** [comment]
How about `TensorInstance` or `TensorType`?

**Comment by yochem** [comment]
No it would not. By using TypeVar: If you call `Tensor.eye()`, the TypeVar type hint understands that `cls` is of type Tensor and the return value of `Tensor.eye()` is Tensor as well. Only if we subclass Tensor as e.g. MyTensor, then `MyTensor.eye()` will return an instance of type MyTensor.

**Comment by prusnak** [comment]
> How about `TensorInstance` or `TensorType`?

the following would be the most standard way and also comprehensible

``` python
TensorType = TypeVar("TensorType", bound=Tensor)
```

**Comment by yochem** [comment]
Went with `TensorType` as proposed by @prusnak.

**Comment by yochem** [comment]
Because other methods are used in the classmethods (e.g. `.reshape()` in `.empty()`), the typevar doesn't work unless we change all methods returning Tensor to return `TensorType` instead. I think this will make things only less clear. I propose to type annotate them for now as `Tensor`, which doesn't support the right typehints for subclasses. When support for 3.8 is dropped, it can be changed to be type hinted as `typing.Self`, as supported by 3.11+.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -31,6 +31,7 @@ def apply(fxn:Type[Function], *x:Tensor, **kwargs) -> Tensor:
 import tinygrad.mlops as mlops
 
 # **** start with two base classes, Tensor and Function ****
+TensorType = TypeVar("TensorType", bound="Tensor")
```
Comment (geohot) [inline tinygrad/tensor.py:34]: Not worth the complexity. Subclassing Tensor isn't really supported

**Comment by geohot** [comment]
Going to close this. Sorry I said it was fine, subclassing Tensor isn't really supported, and now with the custom type it's adding complexity.

PR should be "speed (either emitted or the Python), conceptual simplicity, mlperf training, and bugfixes"

**Comment by yochem** [comment]
And without the custom types and just the staticmethod -> classmethod conversion ([34616296c7d70fb68a6f0d389fd5fd2a6a844718](https://github.com/tinygrad/tinygrad/pull/1716/files/34616296c7d70fb68a6f0d389fd5fd2a6a844718))? I agree that the custom type is adding too much complexity and Tensors being subclassable is not really the goal.

Otherwise sorry for your time! Hope to provide more useful PR's in the future that align with the PR goals :)

**Comment by geohot** [comment]
Probably still not worth it. If there's a real use case for subclassing Tensor we can revisit

**Pull request closed without merging**